config/src/main/java/org/springframework/security/config/ldap/LdapServerBeanDefinitionParser.java

```diff
@@ -142,34 +166,78 @@ private RootBeanDefinition createEmbeddedServer(Element element,
+		String mode = element.getAttribute("mode");
+		RootBeanDefinition ldapContainer = getRootBeanDefinition(mode);
+		ldapContainer.setSource(source);
+		ldapContainer.getConstructorArgumentValues().addGenericArgumentValue(suffix);
 
 		String ldifs = element.getAttribute(ATT_LDIF_FILE);
 		if (!StringUtils.hasText(ldifs)) {
 			ldifs = OPT_DEFAULT_LDIF_FILE;
 		}
 
-		apacheContainer.getConstructorArgumentValues().addGenericArgumentValue(ldifs);
-		apacheContainer.getPropertyValues().addPropertyValue("port", port);
+		ldapContainer.getConstructorArgumentValues().addGenericArgumentValue(ldifs);
+		ldapContainer.getPropertyValues().addPropertyValue("port", port);
 
 		logger.info("Embedded LDAP server bean definition created for URL: " + url);
 
 		if (parserContext.getRegistry()
-				.containsBeanDefinition(BeanIds.EMBEDDED_APACHE_DS)) {
+				.containsBeanDefinition(BeanIds.EMBEDDED_APACHE_DS) ||
+				parserContext.getRegistry().containsBeanDefinition(BeanIds.EMBEDDED_UNBOUNDID)) {
 			parserContext.getReaderContext().error(
 					"Only one embedded server bean is allowed per application context",
 					element);
 		}
 
-		parserContext.getRegistry().registerBeanDefinition(BeanIds.EMBEDDED_APACHE_DS,
-				apacheContainer);
+		EmbeddedLdapServer embeddedLdapServer = resolveEmbeddedLdapServer(mode);
+		if (embeddedLdapServer != null) {
+			parserContext.getRegistry().registerBeanDefinition(embeddedLdapServer.getBeanId(),
+					ldapContainer);
+		}
 
 		return (RootBeanDefinition) contextSource.getBeanDefinition();
 	}
 
+	private RootBeanDefinition getRootBeanDefinition(String mode) {
+		if (StringUtils.hasLength(mode)) {
+			if (isEmbeddedServerEnabled(mode)) {
+				return new RootBeanDefinition(this.embeddedServers.get(mode).getContainerClass(), null, null);
+			}
+		}
+		else {
+			for (Map.Entry<String, EmbeddedLdapServer> entry : this.embeddedServers.entrySet()) {
+				EmbeddedLdapServer ldapServer = entry.getValue();
+				if (ClassUtils.isPresent(ldapServer.getClassName(), getClass().getClassLoader())) {
+					return new RootBeanDefinition(ldapServer.getContainerClass(), null, null);
+				}
+			}
+		}
+		throw new IllegalStateException("Embedded LDAP server is not provided");
+	}
+
+	private boolean isEmbeddedServerEnabled(String mode) {
+		EmbeddedLdapServer server = resolveEmbeddedLdapServer(mode);
+		return server != null;
+	}
+
+	private EmbeddedLdapServer resolveEmbeddedLdapServer(String mode) {
+		if (StringUtils.hasLength(mode)) {
+			if (this.embeddedServers.containsKey(mode) ||
+					ClassUtils.isPresent(this.embeddedServers.get(mode).getClassName(), getClass().getClassLoader())) {
+				return this.embeddedServers.get(mode);
+			}
+		}
+		else {
+			for (Map.Entry<String, EmbeddedLdapServer> entry : this.embeddedServers.entrySet()) {
+				EmbeddedLdapServer ldapServer = entry.getValue();
+				if (ClassUtils.isPresent(ldapServer.getClassName(), getClass().getClassLoader())) {
+					return ldapServer;
+				}
+			}
+		}
+		return null;
```

Comment:
By the time `createEmbeddedServer` is called, aren't we certain that an embedded server needs to be created? When is it ever expected it will be null? When would we expect `isEmbeddedServerEnabled` to be false?